File: Hilos Java/HILOS_METODOS/SemaphoreEjemplo.java
import java.util.concurrent.Semaphore;

public class SemaphoreEjemplo {
    private static Semaphore semaphore = new Semaphore(1);
    
    public static void main(String[] args) {
        Runnable tarea= ()->{
            try {
                System.out.println(Thread.currentThread().getName() + "esperando el acceso...");
                semaphore.acquire();

                System.out.println(Thread.currentThread().getName()+" obtuvo el acceso.");
                Thread.sleep(1000);
            } catch (Exception e) {
                System.out.println("Hilo interrumpido: " +e.getMessage());
            }finally{
                semaphore.release();
                System.out.println(Thread.currentThread().getName()+ " liberó el acceso.");
            }
        };

        Thread[] hilos = new Thread[5];
         
        for(int i = 0; i <5; i++){
            hilos[i] = new Thread(tarea, "Hilo-"+i);
            hilos[i].start();
            try{
                hilos[i].join();
            }catch(InterruptedException e){
                System.out.println("Hilo interrumpido: " + e.getMessage());
            }
        }   
    }
}

File: HILOS_METODOS/SemaphoreEjemplo.py
import threading
import time

class SemaphoreEjemplo:
    semaphore = threading.Semaphore(1)

    @staticmethod
    def tarea():
        try:
            print(f"{threading.current_thread().name} esperando el acceso...")
            SemaphoreEjemplo.semaphore.acquire()

            print(f"{threading.current_thread().name} obtuvo el acceso.")
            time.sleep(1)
        except Exception as e:
            print(f"Hilo interrumpido: {e}")
        finally:
            SemaphoreEjemplo.semaphore.release()
            print(f"{threading.current_thread().name} liberó el acceso.")

if __name__ == "__main__":
    hilos = []
    for i in range(5):
        hilo = threading.Thread(target=SemaphoreEjemplo.tarea, name=f"Hilo-{i}")
        hilos.append(hilo)
        hilo.start()
        hilo.join()
